### PR TITLE
Add Interpolator and InitializeInterpolator.

### PR DESCRIPTION
**Pull request by markscheel**
## Proposed changes

Another PR towards adding parallel interpolator step by step.

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/NumericalAlgorithms/Interpolation/InterpolatedVars.hpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+#include "Domain/BlockId.hpp"
+#include "Domain/ElementId.hpp"
+#include "Domain/Mesh.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+namespace intrp {
+
+/// Data structures holding quantities that are interpolated by
+/// `Interpolator` for use by by `InterpolationTarget`s
```
Comment (kidder) [inline src/NumericalAlgorithms/Interpolation/InterpolatedVars.hpp]: by by

File: src/NumericalAlgorithms/Interpolation/Tags.hpp [hunk 2]
```diff
@@ -26,12 +31,52 @@ struct IndicesOfFilledInterpPoints : db::SimpleTag {
...
 struct TemporalIds : db::SimpleTag {
   using type = std::deque<typename Metavariables::temporal_id>;
   static std::string name() noexcept { return "TemporalIds"; }
 };
 
+/// Volume variables at all `temporal_id`s for all local `Element`s.
+template <typename Metavariables>
+struct VolumeVarsInfo : db::SimpleTag {
+  struct Info {
+    Mesh<3> mesh;
```
Comment (kidder) [inline src/NumericalAlgorithms/Interpolation/Tags.hpp]: template on dim?

File: src/NumericalAlgorithms/Interpolation/Tags.hpp [hunk 3]
```diff
@@ -26,12 +31,52 @@ struct IndicesOfFilledInterpPoints : db::SimpleTag {
...
+    Variables<typename Metavariables::interpolator_source_vars> vars;
+  };
+  using type = std::unordered_map<typename Metavariables::temporal_id,
+                                  std::unordered_map<ElementId<3>, Info>>;
```
Comment (kidder) [inline src/NumericalAlgorithms/Interpolation/Tags.hpp]: template on dim

File: src/NumericalAlgorithms/Interpolation/Tags.hpp [hunk 4]
```diff
@@ -26,12 +31,52 @@ struct IndicesOfFilledInterpPoints : db::SimpleTag {
...
+  static std::string name() noexcept { return "VolumeVarsInfo"; }
+};
+
+namespace holders_detail {
+template <typename Metavariables, typename T>
+struct wrapper {
```
Comment (kidder) [inline src/NumericalAlgorithms/Interpolation/Tags.hpp]: try to use `db::wrap_tags_in` from DataBoxTag.hpp

File: src/NumericalAlgorithms/Interpolation/Tags.hpp [hunk 5]
```diff
@@ -26,12 +31,52 @@ struct IndicesOfFilledInterpPoints : db::SimpleTag {
...
+template <typename Metavariables>
+struct InterpolatedVarsHolders : db::SimpleTag {
+  using type = tuples::tagged_tuple_from_typelist<
+      tmpl::transform<typename Metavariables::interpolation_target_tags,
+                      holders_detail::wrapper<Metavariables, tmpl::_1>>>;
+  static std::string name() noexcept { return "InterpolatedVarsHolders"; }
+};
+
+/// Number of local `Element`s.
+struct NumElements : db::SimpleTag {
```
Comment (kidder) [inline src/NumericalAlgorithms/Interpolation/Tags.hpp]: NumberOfElements

File: src/NumericalAlgorithms/Interpolation/InterpolatedVars.hpp [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+  /// corresponds to only the subset of the points in
+  /// `block_coord_holders` that are contained in local `Element`s.
+  /// Moreover, the number of interpolated points stored in this
+  /// `Info` will change as more `Elements` send data to this
+  /// `Interpolator`, and will be less than or equal to the size of
+  /// `block_coord_holders` even after all `Element`s have sent their
+  /// data (this is because this `Info` lives only on a single core,
+  /// and this core will have access only to the local `Element`s).
+  std::vector<
+      IdPair<domain::BlockId, tnsr::I<double, 3, typename ::Frame::Logical>>>
```
Comment (kidder) [inline src/NumericalAlgorithms/Interpolation/InterpolatedVars.hpp]: 3 -> dim

File: src/NumericalAlgorithms/Interpolation/InterpolatedVars.hpp [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  /// `vars` changes as more `Element`s send data to this `Interpolator`.
+  std::vector<Variables<TagList>> vars{};
+  /// `global_offsets[j][i]` is the index into `block_coord_holders` that
+  /// corresponds to the index `i` of the `DataVector` held in `vars[j]`.
+  /// The size of `global_offsets` changes as more `Element`s
+  /// send data to this `Interpolator`.
+  std::vector<std::vector<size_t>> global_offsets{};
+  /// Holds the `ElementId`s of `Element`s that have already sent their data
+  /// to this `Interpolator`.
+  std::unordered_set<ElementId<3>> element_has_communicated{};
```
Comment (kidder) [inline src/NumericalAlgorithms/Interpolation/InterpolatedVars.hpp]: 3 -> dim

**Review by kidder — CHANGES_REQUESTED** [review]
Also fix line 30 of InitializeInterpolationTarget.hpp (remove the inner of set of ticks)

**Comment by kidder** [comment]
you can squash the changes

**Review by kidder — DISMISSED** [review]

**Review by wthrowe — DISMISSED** [review]

**Review by fmahebert — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
include what you use is sad :(

**Comment by markscheel** [comment]
IWYU is fickle too, but not sad anymore.The changes it wanted this time (adding an include and removing forward declarations) were opposite of what it wanted the last time...

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Comment by kidder** [comment]
travis failures are timeouts

**Pull request merged**